velox/functions/lib/Re2Functions.cpp

```diff
@@ -424,30 +490,101 @@ bool matchSubstringPattern(
       std::string::npos);
 }
 
+// For pattern '___hello___', it is split into three sub-patterns:
+//   - kSingleCharWildcard: ___
+//   - kLiteralString: hello
+//   - kSingleCharWildcard: ___
+//
+// A naive implementation is search the three pattern every time, a better
+// implementation is: skip the length of leading kSingleCharWildcard chars(3
+// here), and search the first kLiteralString string in input to find the
+// position to match.
+bool matchRelaxedSubstringPattern(
+    StringView input,
+    const PatternMetadata& patternMetadata) {
+  if (input.size() < patternMetadata.length()) {
+    return false;
+  }
+
+  // Calculate the necessary search range: [start, end].
+  auto start = patternMetadata.leadingSingleWildcardCharsCount();
+  auto startPatternIndex = start > 0 ? 1 : 0;
+  auto firstFixedString =
+      patternMetadata.subPatterns()[startPatternIndex].pattern.c_str();
+  auto indexInString = start;
+  auto end = input.size() - patternMetadata.restFixedLength() + start;
+
+  while (indexInString <= end) {
+    // Search the firstFixedString to find out where to start match the whole
+    // pattern.
+    auto it = strstr(input.begin() + indexInString, firstFixedString);
+    // The firstFixedString is not found, no need to match anymore.
+    if (!it) {
+      return false;
+    }
+
+    indexInString = it - input.begin();
+    // Check the indexInString is within current input boundary.
+    // Why we need this check? In Velox StringView's data is stored together
+    // in a memory region with no terminating '\0', we need to check the char*
+    // boundary manually.
+    if (indexInString > end) {
+      return false;
+    }
+
+    // Match from the place where find first fixed string.
+    if (matchRelaxedFixedPattern(
+            input,
+            patternMetadata.subPatterns(),
+            patternMetadata.restFixedLength(),
+            indexInString,
+            startPatternIndex)) {
+      return true;
+    } else {
+      // Not match the whole pattern, advance the cursor.
```

Comment:
This can be very slow for some cases like...

'aaaaaaa....aaaaa` LIKE '%_a_b_%'

Reply:
Um... you are right, it is indeed slow for this kind of cases, will move kRelaxedSubstring optimization into separate PR.